### PR TITLE
Support ACME for certificate provisioning

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

File: synapse/handlers/acme.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 New Vector Ltd
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/handlers/acme.py]: it's still 2018 in my heart

File: synapse/app/homeserver.py [hunk 2]
```diff
@@ -361,9 +367,37 @@ def setup(config_options):
...
+    if hs.config.acme_enabled:
+        acme = hs.get_acme_handler()
+        acme.start_listening()
+    else:
+        hs.start_listening()
 
     def start():
+        if hs.config.acme_enabled:
+            is_valid_cert = acme.is_disk_cert_valid()
+            if not is_valid_cert:
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: Yeah, that was one of the next little thinkies that I have to tackle. I was going to do it in the "enable cert reloads without restarting", since it  means I have to _stop_ things listening and then start them again, which I think is making this a bit too big of a PR.

File: synapse/handlers/_base.py [hunk 3]
```diff
@@ -46,6 +46,7 @@ def __init__(self, hs):
         self.distributor = hs.get_distributor()
         self.ratelimiter = hs.get_ratelimiter()
         self.clock = hs.get_clock()
+        self.reactor = hs.get_reactor()
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/handlers/_base.py]: sure

File: synapse/app/homeserver.py [hunk 4]
```diff
@@ -361,9 +367,37 @@ def setup(config_options):
...
+
+            def _load_context_factories(_):
+                hs.tls_server_context_factory = context_factory.ServerContextFactory(
+                    config
+                )
+                hs.tls_client_options_factory = context_factory.ClientTLSOptionsFactory(
+                    config
+                )
+
+            d.addCallback(lambda _: hs.config._read_certificate())
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: sure, good idea.

File: synapse/handlers/acme.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            tls_private_key = crypto.load_privatekey(crypto.FILETYPE_PEM, cert_chain)
+            with open(self.hs.config.tls_private_key_file, "wb") as private_key_file:
+                private_key_pem = crypto.dump_privatekey(
+                    crypto.FILETYPE_PEM, tls_private_key
+                )
+                private_key_file.write(private_key_pem)
+
+            cert = crypto.load_certificate(crypto.FILETYPE_PEM, cert_chain)
+            with open(self.hs.config.tls_certificate_file, "wb") as certificate_file:
+                cert_pem = crypto.dump_certificate(crypto.FILETYPE_PEM, cert)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/handlers/acme.py]: done

File: synapse/app/homeserver.py [hunk 6]
```diff
@@ -361,9 +367,37 @@ def setup(config_options):
...
+    # have a certificate, first.
+    if hs.config.acme_enabled:
+        acme = hs.get_acme_handler()
+        acme.start_listening()
+    else:
+        hs.start_listening()
 
     def start():
+        if hs.config.acme_enabled:
+            is_valid_cert = acme.is_disk_cert_valid()
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: yeah i simplified it

File: synapse/app/homeserver.py [hunk 7]
```diff
@@ -361,9 +367,37 @@ def setup(config_options):
...
+            if not is_valid_cert:
+                d = acme._issuer._ensure_registered()
+                d.addCallback(
+                    lambda _: acme.provision_certificate(hs.config.server_name)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: done

File: synapse/config/tls.py [hunk 8]
```diff
@@ -118,6 +142,17 @@ def default_config(self, config_dir_path, server_name, **kwargs):
...
+        ## Support for ACME certificate auto-provisioning.
+        # acme:
+        #    enabled: false
+        ##   ACME path. Default: https://acme-staging.api.letsencrypt.org/directory
+        #    url: 'https://acme-v01.api.letsencrypt.org/directory'
+        ##   Port number (to listen for the HTTP-01 challenge).
+        ##   Using port 80 requires utilising something like authbind, or proxying to it.
+        #    port: 8449
+        ##   Hosts to bind to, comma separated.
+        #    host: '127.0.0.1'
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/config/tls.py]: done

File: synapse/app/homeserver.py [hunk 9]
```diff
@@ -324,8 +325,13 @@ def setup(config_options):
 
     events.USE_FROZEN_DICTS = config.use_frozen_dicts
 
-    tls_server_context_factory = context_factory.ServerContextFactory(config)
-    tls_client_options_factory = context_factory.ClientTLSOptionsFactory(config)
+    # These will be loaded in later, once we have provisioned keys via ACME.
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: done

File: synapse/app/homeserver.py [hunk 10]
```diff
@@ -361,9 +367,37 @@ def setup(config_options):
...
+        acme = hs.get_acme_handler()
+        acme.start_listening()
+    else:
+        hs.start_listening()
 
     def start():
+        if hs.config.acme_enabled:
+            is_valid_cert = acme.is_disk_cert_valid()
+            if not is_valid_cert:
+                d = acme._issuer._ensure_registered()
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: not really, but i encapsulated it a little better

File: synapse/app/homeserver.py [hunk 11]
```diff
@@ -361,9 +357,43 @@ def setup(config_options):
     logger.info("Database prepared in %s.", config.database_config['name'])
 
     hs.setup()
-    hs.start_listening()
```
Comment (richvdh) [inline synapse/app/homeserver.py]: this generally looks much saner now.

File: synapse/app/homeserver.py [hunk 12]
```diff
@@ -361,9 +357,43 @@ def setup(config_options):
...
+            # certificate exists), or the days remaining number it returns is
+            # less than our re-registration threshold.
+            if (cert_days_remaining is None) or (
+                not cert_days_remaining > hs.config.acme_reprovision_threshold
+            ):
+                yield acme.provision_certificate()
+
+        # Read the certificate from disk and build the context factories for
+        # TLS.
+        hs.config._read_certificate()
```
Comment (richvdh) [inline synapse/app/homeserver.py]: please can this lose its `_` if it's going to be public

File: synapse/app/homeserver.py [hunk 13]
```diff
@@ -361,9 +357,43 @@ def setup(config_options):
...
+        if hs.config.acme_enabled:
+            # If ACME is enabled, we might need to provision a certificate
+            # before starting.
+            acme = hs.get_acme_handler()
+
+            # Start up the webservices which we will respond to ACME challenges
+            # with.
+            try:
+                yield acme.start_listening()
+            except defer.FirstError as e:
```
Comment (richvdh) [inline synapse/app/homeserver.py]: it feels fragile to tie this to FirstError. should it not extend to any Exception? (and indeed, should it not apply to the whole of `start` ?)

File: synapse/app/homeserver.py [hunk 14]
```diff
@@ -361,9 +357,43 @@ def setup(config_options):
...
+                # We couldn't listen on a port, for some reason. Print the
+                # traceback to stderr and bail.
+                e.subFailure.printTraceback(sys.stderr)
+                sys.exit(1)
+
+            # We want to reprovision if is_valid_cert is None (meaning no
```
Comment (richvdh) [inline synapse/app/homeserver.py]:
```suggestion
            # We want to reprovision if cert_days_remaining is None (meaning no
```

File: synapse/config/tls.py [hunk 15]
```diff
@@ -15,66 +15,99 @@
...
+        )
 
-        self.tls_dh_params_path = self.check_file(
-            config.get("tls_dh_params_path"), "tls_dh_params"
+        self.tls_certificate = None
+        self.tls_private_key = None
+
+        # If we are using ACME, do not read the certificate yet. That will be
+        # done later, and will trigger the reading code.
+        if not self.acme_enabled:
```
Comment (richvdh) [inline synapse/config/tls.py]: this seems redundant now? If there's a good reason to do it anyway, can you comment to say so?

File: synapse/config/tls.py [hunk 16]
```diff
@@ -15,66 +15,99 @@
...
+            self._read_certificate()
+
+    def is_disk_cert_valid(self):
+        """
+        Is the certificate we have on disk valid, and if so, for how long?
+        """
```
Comment (richvdh) [inline synapse/config/tls.py]: please can you document the return types?

File: synapse/config/tls.py [hunk 17]
```diff
@@ -15,66 +15,99 @@
...
+        try:
+            with open(self.tls_certificate_file, 'rb') as f:
+                cert_pem = f.read()
+            tls_certificate = crypto.load_certificate(crypto.FILETYPE_PEM, cert_pem)
```
Comment (richvdh) [inline synapse/config/tls.py]: this probably wants to come out of the `try` block?

File: synapse/config/tls.py [hunk 18]
```diff
@@ -15,66 +15,99 @@
...
+        except IOError:
+            return None
```
Comment (richvdh) [inline synapse/config/tls.py]: can we distinguish between "doesn't exist" and "couldn't read it", and only return None in the former case?

an INFO line might be useful too.

File: synapse/config/tls.py [hunk 19]
```diff
@@ -85,9 +118,6 @@ def default_config(self, config_dir_path, server_name, **kwargs):
         # PEM encoded private key for TLS
         tls_private_key_path: "%(tls_private_key_path)s"
 
-        # PEM dh parameters for ephemeral keys
-        tls_dh_params_path: "%(tls_dh_params_path)s"
```
Comment (richvdh) [inline synapse/config/tls.py]: fully on board with killing this off; however can you grep for it appearing elsewhere and make sure that we kill it off properly? (it's in the README for one, and we might want to think about getting rid of it from the debian package and docker image)

it kinda would have been nice to pull it out to a separate PR tbh... or perhaps just do the cleanup in a separate PR now we've got this far.

File: synapse/python_dependencies.py [hunk 20]
```diff
@@ -72,6 +72,7 @@
     # ConsentResource uses select_autoescape, which arrived in jinja 2.9
     "resources.consent": ["Jinja2>=2.9"],
 
+    "acme": ["txacme>=0.9.2"],
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/python_dependencies.py]: still hoping for a comment on what "acme" means.

[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/python_dependencies.py]: still hoping for a comment on what "acme" means.

File: synapse/app/homeserver.py (hunk 13 above)
Comment (hawkowl) [reply synapse/app/homeserver.py]: Yeahhhh I guess catching the entirety (and then special casing FirstError, which can only come from the DeferredList to provide a nicer log message) will be better.

Not sure about making the entirety of `start` in a block. I guess we ought to, since there's now more in it?

**Comment by hawkowl** [comment]
> Something I don't quite understand here. I thought you needed an account key to get certs from LE, and that you were supposed to reuse the same account key for subsequent renewals. I'm not seeing anything about that here... what am I missing?

@richvdh  This is because it registers and uses a consistent `client.key`. See https://github.com/matrix-org/synapse/pull/4384/files#diff-149d848a6b79ba9483b2e5e40ee029feR82 -- txacme terms it as a client key, not an account key.

File: synapse/app/homeserver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/app/homeserver.py]: as it stands, it will renew if there's less than 10 days left, so it shouldn't boot with a cert which will immediately expire. Of course, that's not a permanent solution, but it gives people who want to immediately use the feature in the meantime some headroom.

**Comment by hawkowl** [comment]
@richvdh I'll pull the `dh_params` out, as it does need a bit of cleanup in other areas + documentation, and could really use its own newsfile. The changing of the cipher string shouldn't result in anything bad (as what's there should equate to what I've added) but will be more consistent across different versions of OpenSSL and is understandable without reading a man page (which I still had trouble deciphering), but is worth calling out anyway.

File: synapse/app/homeserver.py (hunk 13 above)
Comment (richvdh) [reply synapse/app/homeserver.py]: yeah I guess we ought to, really. Mostly it felt odd to add a special-case for this one line.

File: synapse/config/tls.py [hunk 21]
```diff
@@ -15,66 +15,99 @@
...
-        self.tls_certificate = self.read_tls_certificate(
-            config.get("tls_certificate_path")
+
+        acme_config = config.get("acme", {})
+        self.acme_enabled = acme_config.get("enabled", False)
+        self.acme_url = acme_config.get(
+            "url", "https://acme-staging.api.letsencrypt.org/directory"
         )
-        self.tls_certificate_file = config.get("tls_certificate_path")
+        self.acme_client_key = config.get("client_key", self.config_dir_path)
```
Comment (richvdh) [inline synapse/config/tls.py]: please could this be called `[acme_]client_key_*path*`, since it's a filename rather than the key itself?

Comment (richvdh) [inline synapse/config/tls.py]: also, it probably needs a call to `self.abspath`, which is the convention for filenames in the config file (synapse does a `chdir` after loading the config).

Comment (richvdh) [inline synapse/config/tls.py]: also, the default looks bogus.

File: synapse/config/tls.py [hunk 22]
```diff
@@ -118,7 +148,22 @@ def default_config(self, config_dir_path, server_name, **kwargs):
         #
         tls_fingerprints: []
         # tls_fingerprints: [{"sha256": "<base64_encoded_sha256_fingerprint>"}]
-        """ % locals()
+
+        ## Support for ACME certificate auto-provisioning.
+        # acme:
```
Comment (richvdh) [inline synapse/config/tls.py]: `acme_client_key` is missing from this list. Current convention is to list all the options here, even if they aren't really expected to be used much, since there's no other reference for the config options.

File: synapse/app/homeserver.py (hunk 12 above)
Comment (hawkowl) [reply synapse/app/homeserver.py]: sure

File: synapse/config/tls.py (hunk 17 above)
Comment (hawkowl) [reply synapse/config/tls.py]: yeah

File: synapse/config/tls.py (hunk 18 above)
Comment (hawkowl) [reply synapse/config/tls.py]: hmm. I am not entirely sure about the difference to the code. is it that we would want to raise and bail out if we couldn't read the certificate?

File: synapse/app/homeserver.py (hunk 13 above)
Comment (hawkowl) [reply synapse/app/homeserver.py]: done

File: synapse/config/tls.py (hunk 15 above)
Comment (hawkowl) [reply synapse/config/tls.py]: yeah it's redundant. removed.

File: synapse/config/tls.py (hunk 16 above)
Comment (hawkowl) [reply synapse/config/tls.py]: done

File: synapse/config/tls.py (hunk 18 above)
Comment (hawkowl) [reply synapse/config/tls.py]: now: returns None if it doesn't exist, bails out if it can't read it or can't parse it.

File: synapse/config/tls.py [hunk 23]
```diff
@@ -13,60 +13,110 @@
...
-        self.tls_certificate_file = config.get("tls_certificate_path")
 
+        acme_config = config.get("acme", {})
+        self.acme_enabled = acme_config.get("enabled", False)
+        self.acme_url = acme_config.get(
+            "url", "https://acme-staging.api.letsencrypt.org/directory"
+        )
+        self.acme_port = acme_config.get("port", 8449)
+        self.acme_bind_addresses = acme_config.get("bind_addresses", ["127.0.0.1"])
+        self.acme_reprovision_threshold = acme_config.get("reprovision_threshold", 10)
```
Comment (michaelkaye) [inline synapse/config/tls.py]: LetsEncrypt as the main cert provider via ACME suggest renewal after 60 days (at 30 days to go). They currently send emails to warn of un-renewed certs at 20 and 10 days, so this may want to be about 30 days as a default.

File: synapse/config/tls.py [hunk 24]
```diff
@@ -13,60 +13,110 @@
...
     def read_config(self, config):
-        self.tls_certificate = self.read_tls_certificate(
-            config.get("tls_certificate_path")
-        )
-        self.tls_certificate_file = config.get("tls_certificate_path")
 
+        acme_config = config.get("acme", {})
+        self.acme_enabled = acme_config.get("enabled", False)
+        self.acme_url = acme_config.get(
+            "url", "https://acme-staging.api.letsencrypt.org/directory"
```
Comment (michaelkaye) [inline synapse/config/tls.py]: Don't forget to set the default to be the live endpoint

Comment (michaelkaye) [reply synapse/config/tls.py]: Also, do we want to consider using the v2 API now , if it's a simple change in our use of the library, to prevent us having to move to v2 when they decommission (timeline unspecified, so if it's a lot of work it is definitelyworth punting down the road)

File: synapse/app/homeserver.py [hunk 25]
```diff
@@ -361,12 +358,53 @@ def setup(config_options):
...
+            if hs.config.acme_enabled:
+                # If ACME is enabled, we might need to provision a certificate
+                # before starting.
+                acme = hs.get_acme_handler()
+
+                # Start up the webservices which we will respond to ACME
+                # challenges with.
+                yield acme.start_listening()
+
+                # We want to reprovision if is_valid_cert is None (meaning no
```
Comment (richvdh) [inline synapse/app/homeserver.py]: still:

```suggestion
                # We want to reprovision if cert_days_remaining is None (meaning no
```

File: synapse/config/tls.py (hunk 24 above)
Comment (richvdh) [reply synapse/config/tls.py]: that feels like a lower risk then them wondering why their cert doesn't work, tbh. The fact that the whole thing is disabled by default is enough of a safetynet imho.

Comment (hawkowl) [reply synapse/config/tls.py]: hmm, okay
